python/lsst/daf/butler/core/dimensions/_graph.py

```diff
@@ -229,46 +231,23 @@ def __str__(self) -> str:
     def __repr__(self) -> str:
         return f"DimensionGraph({str(self)})"
 
-    @classmethod
-    def decode(cls, encoded: bytes, *, universe: DimensionUniverse) -> DimensionGraph:
-        """Construct a `DimensionGraph` from its encoded representation.
-
-        Parameters
-        ----------
-        encoded : `bytes`
-            Byte string produced by `DimensionGraph.encode`.
-        universe : `DimensionUniverse`
-            Universe the new graph is a part of.  Must have the same dimensions
-            as the original universe.
-
-        Returns
-        -------
-        graph : `DimensionGraph`
-            A new (or possibly cached) `DimensionGraph` instance matching the
-            given encoding.
-        """
-        dimensions = []
-        mask = int.from_bytes(encoded, "big")
-        for dimension in universe.getStaticDimensions():
-            index = universe.getDimensionIndex(dimension.name)
-            if mask & (1 << index):
-                dimensions.append(dimension)
-        return cls(universe, dimensions=dimensions, conform=False)
+    DIGEST_SIZE: ClassVar[int] = 32
+    """String size of the hexidecimal hashes returned by `digest`.
+    """
 
-    def encode(self) -> bytes:
-        """Encode a `DimensionGraph` into a byte string.
+    def digest(self) -> str:
+        """Compute a secure hash of this DimensionGraph, suitable for use as a
+        unique ID for it.
 
         Returns
         -------
-        encoded : `bytes`
-            Encoded representation of the graph.  Length is guaranteed to be
-            equal to `DimensionUniverse.getEncodeLength`.
+        digest : `str`
+            Secure hash of ``self``, in hexadecimal.
         """
-        mask = 0
-        for dimension in self.dimensions:
-            index = self.universe.getDimensionIndex(dimension.name)
-            mask |= (1 << index)
-        return mask.to_bytes(self.universe.getEncodeLength(), byteorder="big")
+        hasher = hashlib.blake2b(digest_size=self.DIGEST_SIZE//2)
+        for name in self.required.names:
+            hasher.update(name.encode("ascii"))
+        return hasher.hexdigest()
```

Comment:
I do not trust hashes in general due to potential collisions. In this case collisions are possible for cases like `["a", "b"]` vs just `["ab"]`. It may help adding a non-word separator character (e.g. TAB) to avoid this issue.

Reply:
I'll at least add that separator; good idea.

I also considered just using an autoincrement integer for the ID (so there would be no `DimensionGraph` method to get it).  The problem with that is that we'd have to load essentially the entire table from the database in order to see if a particular graph already has an ID, though that's more about code complexity than performance because the table should always be small.  I also liked the idea of making the ID deterministic across different data repositories, but I don't have a concrete reason why that needs to be the case.  Do you think an autoincrement ID would be better here?

Reply:
Deterministic ID would be nice to have but I guess it's hard to construct or make it stable. If we do not need this ID outside database relation context then autoincrement is probably a better match for that. That table is small so it just a single query/round-trip to database, so I would not worry about performance, and code complexity should not be terrible.

Reply:
I've added another commit that uses autoincrement integer IDs instead; please take a look.  Once the review is finished I'll squash commits and probably just remove `DimensionGraph.digest`.